src/gstring.rs

```diff
@@ -0,0 +1,445 @@
+    fn borrow(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl Ord for GString {
+    fn cmp(&self, other: &GString) -> Ordering {
+        self.as_str().cmp(other.as_str())
+    }
+}
+
+impl PartialOrd for GString {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for GString {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<GString> for String {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<str> for GString {
+    fn eq(&self, other: &str) -> bool {
+        self.as_str() == other
+    }
+}
+
+impl<'a> PartialEq<&'a str> for GString {
+    fn eq(&self, other: &&'a str) -> bool {
+        self.as_str() == *other
+    }
+}
+
+impl<'a> PartialEq<GString> for &'a str {
+    fn eq(&self, other: &GString) -> bool {
+        *self == other.as_str()
+    }
+}
+
+impl PartialEq<String> for GString {
+    fn eq(&self, other: &String) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<GString> for str {
+    fn eq(&self, other: &GString) -> bool {
+        self == other.as_str()
+    }
+}
+
+impl PartialOrd<GString> for String {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(&String::from(other.as_str())))
+    }
+}
+
+impl PartialOrd<String> for GString {
+    fn partial_cmp(&self, other: &String) -> Option<Ordering> {
+        Some(self.as_str().cmp(other.as_str()))
+    }
+}
+
+impl PartialOrd<GString> for str {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl PartialOrd<str> for GString {
+    fn partial_cmp(&self, other: &str) -> Option<Ordering> {
+        Some(self.as_str().cmp(other))
+    }
+}
+
+impl Eq for GString {}
+
+impl AsRef<str> for GString {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl AsRef<OsStr> for GString {
+    fn as_ref(&self) -> &OsStr {
+        OsStr::new(self.as_str())
+    }
+}
+
+impl Deref for GString {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl From<GString> for String {
+    #[inline]
+    fn from(s: GString) -> Self {
+        String::from(s.as_str())
+    }
+}
+
+impl From<GString> for Box<str> {
+    #[inline]
+    fn from(s: GString) -> Self {
+        let st: String = s.into();
+        st.into_boxed_str()
+    }
+}
+
+impl From<String> for GString {
+    #[inline]
+    fn from(s: String) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<Box<str>> for GString {
+    #[inline]
+    fn from(s: Box<str>) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<&'static str> for GString {
+    #[inline]
+    fn from(s: &'static str) -> Self {
+        GString::Borrowed(s.as_ptr() as *const c_char, s.len())
+    }
+}
+
+impl From<Vec<u8>> for GString {
```

Comment:
Can this be made generic as `impl<T: Into<Vec<u8>>> From<T> for GString {}` or does that break things?

Reply:
I tried this but then the `FromGlibPtrNone` impls can't rely on `CStr`:
```
error[E0277]: the trait bound `std::vec::Vec<u8>: std::convert::From<&std::ffi::CStr>` is not satisfied                                                                                                            
   --> src/gstring.rs:270:14                                                                                                                                                                                       
    |                                                                                                                                                                                                              
270 |         cstr.into()                                                                                                                                                                                          
    |              ^^^^ the trait `std::convert::From<&std::ffi::CStr>` is not implemented for `std::vec::Vec<u8>`                                                                                                 
    |                                                                                                                                                                                                              
    = help: the following implementations were found:                                                                                                                                                              
              <std::vec::Vec<T> as std::convert::From<std::boxed::Box<[T]>>>                                                                                                                                       
              <std::vec::Vec<u8> as std::convert::From<&'a str>>                                                                                                                                                   
              <std::vec::Vec<u8> as std::convert::From<std::ffi::CString>>                                                                                                                                         
              <std::vec::Vec<u8> as std::convert::From<std::string::String>>                                                                                                                                       
            and 5 others                                                                                                                                                                                           
    = note: required because of the requirements on the impl of `std::convert::Into<std::vec::Vec<u8>>` for `&std::ffi::CStr`                                                                                      
    = note: required because of the requirements on the impl of `std::convert::From<&std::ffi::CStr>` for `gstring::GString`                                                                                       
    = note: required because of the requirements on the impl of `std::convert::Into<gstring::GString>` for `&std::ffi::CStr`        
```

Reply:
I could use `CString::from_raw` but its documentation doesn't recommend it :)

Reply:
Ok then :)